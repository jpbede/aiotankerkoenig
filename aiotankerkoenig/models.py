"""Models for the aiotankerkoenig client."""
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Self

from mashumaro import field_options
from mashumaro.mixins.orjson import DataClassORJSONMixin

from .const import Status
from .exceptions import TankerkoenigError, TankerkoenigInvalidKeyError


@dataclass(frozen=True, slots=True, kw_only=True)
class TankerkoenigResponse(DataClassORJSONMixin):
    """Base class for all responses."""

    @classmethod
    def __pre_deserialize__(
        cls: type[Self],
        d: dict[Any, Any],
    ) -> dict[Any, Any]:
        """Raise when response was unexpected."""
        if d.get("ok"):
            return d

        message = d.get("message", "")
        if any(x in message.lower() for x in ("api-key", "apikey")):
            msg = "tankerkoenig.de API responded with an invalid key error"
            raise TankerkoenigInvalidKeyError(
                msg,
                {"response": d},
            )

        msg = f"tankerkoenig.de API responded with an error: {message}"
        raise TankerkoenigError(
            msg,
            {"response": d.pop("message")},
        )


@dataclass(frozen=True, slots=True, kw_only=True)
class StationListResponse(TankerkoenigResponse):
    """Class representing a station list response."""

    stations: list[Station]


@dataclass(frozen=True, slots=True, kw_only=True)
class StationDetailResponse(TankerkoenigResponse):
    """Class representing a station response."""

    station: Station


@dataclass(frozen=True, slots=True, kw_only=True)
class PriceInfoResponse(TankerkoenigResponse):
    """Class representing a station price info response."""

    prices: dict[str, PriceInfo]


@dataclass(frozen=True, slots=True, kw_only=True)
class OpeningTime:
    """Class representing a station opening time."""

    end: str
    start: str
    text: str


@dataclass(frozen=True, slots=True, kw_only=True)
class Station:
    """Class representing a station."""

    id: str
    name: str
    house_number: str = field(metadata=field_options(alias="houseNumber"))
    place: str
    post_code: int = field(metadata=field_options(alias="postCode"))
    street: str
    lat: float
    lng: float

    # only available in detail view
    brand: str | None = None
    diesel: float | None = None
    e10: float | None = None
    e5: float | None = None
    is_open: bool | None = field(metadata=field_options(alias="isOpen"), default=None)
    overrides: Any | None = None
    whole_day: bool | None = field(
        metadata=field_options(alias="wholeDay"),
        default=None,
    )
    state: str | None = None
    opening_times: list[OpeningTime] | None = field(
        metadata=field_options(alias="opening_times"),
        default=None,
    )
    distance: float | None = field(metadata=field_options(alias="dist"), default=None)


@dataclass(frozen=True, slots=True, kw_only=True)
class PriceInfo:
    """Class representing a station price info."""

    status: Status = Status.UNKNOWN
    e5: float | None = None
    e10: float | None = None
    diesel: float | None = None
