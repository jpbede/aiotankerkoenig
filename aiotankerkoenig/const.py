"""Constants for the aiotankerkoenig."""
from enum import StrEnum


class GasType(StrEnum):
    """Gas type."""

    ALL = "all"
    DIESEL = "diesel"
    E5 = "e5"
    E10 = "e10"


class Sort(StrEnum):
    """Sort type."""

    DISTANCE = "dist"
    PRICE = "price"
    TIME = "time"


class Status(StrEnum):
    """Status type."""

    OPEN = "open"
    CLOSED = "closed"
    UNKNOWN = "unknown"
