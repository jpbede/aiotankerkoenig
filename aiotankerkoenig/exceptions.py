"""Exceptions for the aiotankerkoenig client."""
from __future__ import annotations


class TankerkoenigError(Exception):
    """Base exception class for all exceptions raised by this library."""


class TankerkoenigConnectionError(TankerkoenigError):
    """Raised when a connection error occurs."""


class TankerkoenigConnectionTimeoutError(TankerkoenigConnectionError):
    """Raised when a connection times out."""


class TankerkoenigInvalidKeyError(TankerkoenigError):
    """Raised when the API key is invalid."""


class TankerkoenigRateLimitError(TankerkoenigError):
    """Raised when the API rate limit is reached."""
