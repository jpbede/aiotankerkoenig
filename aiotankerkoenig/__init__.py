"""Tankerkoenig API client."""
from __future__ import annotations

from .aiotankerkoenig import Tankerkoenig
from .const import GasType, Sort, Status
from .exceptions import (
    TankerkoenigConnectionError,
    TankerkoenigConnectionTimeoutError,
    TankerkoenigError,
    TankerkoenigInvalidKeyError,
    TankerkoenigRateLimitError,
)
from .models import PriceInfo, Station

__all__ = [
    "Tankerkoenig",
    "TankerkoenigError",
    "TankerkoenigConnectionError",
    "TankerkoenigInvalidKeyError",
    "TankerkoenigConnectionTimeoutError",
    "TankerkoenigRateLimitError",
    "GasType",
    "Sort",
    "Status",
    "PriceInfo",
    "Station",
]
