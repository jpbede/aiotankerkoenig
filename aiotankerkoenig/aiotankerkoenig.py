"""Tankerkoenig API client."""
from __future__ import annotations

from dataclasses import dataclass
from importlib import metadata
import socket
from typing import Any, Self

from aiohttp import ClientError, ClientResponseError, ClientSession
from yarl import URL

from .const import GasType, Sort
from .exceptions import (
    TankerkoenigConnectionError,
    TankerkoenigConnectionTimeoutError,
    TankerkoenigError,
    TankerkoenigRateLimitError,
)
from .models import (
    PriceInfo,
    PriceInfoResponse,
    Station,
    StationDetailResponse,
    StationListResponse,
)

VERSION = metadata.version(__package__)


@dataclass
class Tankerkoenig:
    """Tankerkoenig API client."""

    api_key: str
    session: ClientSession | None = None

    _close_session: bool = False

    async def _request(self, path: str, params: dict[str, Any]) -> str:
        """Handle request to tankerkoenig.de API."""
        url = URL.build(
            scheme="https",
            host="creativecommons.tankerkoenig.de",
            path=path,
            query={"apikey": self.api_key, **params},
        )

        headers = {
            "Accept": "application/json",
        }

        if self.session is None:
            self.session = ClientSession()
            self._close_session = True
            headers.update(
                {
                    "User-Agent": f"aiotankerkoenig/{VERSION}",
                },
            )

        try:
            async with self.session.get(url, headers=headers) as response:
                response.raise_for_status()
                content_type = response.headers.get("Content-Type", "")
                text = await response.text()
        except TimeoutError as exception:
            msg = "Timeout occurred while connecting to tankerkoenig.de API"
            raise TankerkoenigConnectionTimeoutError(
                msg,
            ) from exception
        except (
            ClientError,
            socket.gaierror,
        ) as exception:
            if isinstance(exception, ClientResponseError) and exception.code == 503:
                raise TankerkoenigRateLimitError from exception
            msg = "Error occurred while communicating with the tankerkoenig.de API"
            raise TankerkoenigConnectionError(msg) from exception

        if "application/json" not in content_type:
            msg = "Unexpected content type from tankerkoenig.de API"
            raise TankerkoenigError(
                msg,
                {"Content-Type": content_type, "response": text},
            )

        return text

    async def nearby_stations(
        self,
        coordinates: tuple[float, float],
        radius: int,
        gas_type: GasType,
        sort: Sort,
    ) -> list[Station]:
        """Get nearby stations."""
        result = await self._request(
            path="/json/list.php",
            params={
                "lat": coordinates[0],
                "lng": coordinates[1],
                "rad": radius,
                "type": gas_type,
                "sort": sort,
            },
        )
        return StationListResponse.from_json(result).stations

    async def station_details(
        self,
        station_id: str,
    ) -> Station:
        """Get station details."""
        result = await self._request(
            path="/json/detail.php",
            params={
                "id": station_id,
            },
        )
        return StationDetailResponse.from_json(result).station

    async def prices(
        self,
        station_ids: list[str],
    ) -> dict[str, PriceInfo]:
        """Get station details."""
        result = await self._request(
            path="/json/prices.php",
            params={
                "ids": ",".join(station_ids),
            },
        )
        return PriceInfoResponse.from_json(result).prices

    async def close(self) -> None:
        """Close open client session."""
        if self.session and self._close_session:
            await self.session.close()

    async def __aenter__(self) -> Self:
        """Async enter."""
        return self

    async def __aexit__(self, *_exc_info: object) -> None:
        """Async exit.    _exc_info: Exec type."""
        await self.close()
