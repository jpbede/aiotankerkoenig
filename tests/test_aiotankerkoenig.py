"""Test the tankerkoenig API client."""
import asyncio
from typing import Any

import aiohttp
from aioresponses import CallbackResult, aioresponses
import pytest
from syrupy import SnapshotAssertion

from aiotankerkoenig import (
    GasType,
    Sort,
    Tankerkoenig,
    TankerkoenigConnectionError,
    TankerkoenigConnectionTimeoutError,
    TankerkoenigError,
    TankerkoenigInvalidKeyError,
)
from tests import load_fixture

TANKERKOENIG_ENDPOINT = "https://creativecommons.tankerkoenig.de"


async def test_timeout(
    responses: aioresponses,
) -> None:
    """Test request timeout."""

    async def response_handler(_: str, **_kwargs: Any) -> CallbackResult:
        """Response handler for this test."""
        await asyncio.sleep(2)
        return CallbackResult(body="Good morning!")

    responses.get(
        f"{TANKERKOENIG_ENDPOINT}/json/detail.php?apikey=abc123&id=1",
        callback=response_handler,
    )
    async with Tankerkoenig(
        api_key="abc123",
        request_timeout=1,
    ) as tankerkoenig_client:
        with pytest.raises(TankerkoenigConnectionTimeoutError):
            await tankerkoenig_client.station_details(station_id="1")


async def test_unexpected_server_response(
    responses: aioresponses,
    tankerkoenig_client: Tankerkoenig,
) -> None:
    """Test handling unexpected response."""
    responses.get(
        f"{TANKERKOENIG_ENDPOINT}/json/detail.php?apikey=abc123&id=1",
        status=200,
        headers={"Content-Type": "plain/text"},
        body="Yes",
    )
    with pytest.raises(TankerkoenigError):
        await tankerkoenig_client.station_details(station_id="1")


async def test_non_200_response(
    responses: aioresponses,
    tankerkoenig_client: Tankerkoenig,
) -> None:
    """Test handling unexpected response."""
    responses.get(
        f"{TANKERKOENIG_ENDPOINT}/json/detail.php?apikey=abc123&id=1",
        status=500,
        headers={"Content-Type": "plain/text"},
        body="Yes",
    )
    with pytest.raises(TankerkoenigConnectionError):
        await tankerkoenig_client.station_details(station_id="1")


@pytest.mark.parametrize(
    "message",
    [
        "API-Key existiert nicht",
        "APIKey existiert nicht.",
    ],
)
async def test_invalid_apikey_response(
    responses: aioresponses,
    tankerkoenig_client: Tankerkoenig,
    message: str,
) -> None:
    """Test handling of invalid api key response."""
    responses.get(
        f"{TANKERKOENIG_ENDPOINT}/json/detail.php?apikey=abc123&id=1",
        status=200,
        body='{"status": "error", "ok": false, "message": "' + message + '"}',
    )
    with pytest.raises(TankerkoenigInvalidKeyError):
        await tankerkoenig_client.station_details(station_id="1")


async def test_general_error_response(
    responses: aioresponses,
    tankerkoenig_client: Tankerkoenig,
) -> None:
    """Test handling of a general error response."""
    responses.get(
        f"{TANKERKOENIG_ENDPOINT}/json/detail.php?apikey=abc123&id=1",
        status=200,
        body='{"status": "error", "ok": false, "message": "Idk what happened!"}',
    )
    with pytest.raises(TankerkoenigError):
        await tankerkoenig_client.station_details(station_id="1")


async def test_putting_in_own_session(
    responses: aioresponses,
) -> None:
    """Test putting in own session."""
    responses.get(
        f"{TANKERKOENIG_ENDPOINT}/json/detail.php?apikey=abc123&id=1",
        status=200,
        body=load_fixture("detail.json"),
    )
    async with aiohttp.ClientSession() as session:
        tankerkoenig = Tankerkoenig(api_key="abc123", session=session)
        await tankerkoenig.station_details(station_id="1")
        assert tankerkoenig.session is not None
        assert not tankerkoenig.session.closed
        await tankerkoenig.close()
        assert not tankerkoenig.session.closed


async def test_creating_own_session(
    responses: aioresponses,
) -> None:
    """Test creating own session."""
    responses.get(
        f"{TANKERKOENIG_ENDPOINT}/json/detail.php?apikey=abc123&id=1",
        status=200,
        body=load_fixture("detail.json"),
    )
    tankerkoenig = Tankerkoenig(api_key="abc123")
    await tankerkoenig.station_details(station_id="1")
    assert tankerkoenig.session is not None
    assert not tankerkoenig.session.closed
    await tankerkoenig.close()
    assert tankerkoenig.session.closed


async def test_station_details(
    responses: aioresponses,
    tankerkoenig_client: Tankerkoenig,
    snapshot: SnapshotAssertion,
) -> None:
    """Test retrieving station details."""
    responses.get(
        f"{TANKERKOENIG_ENDPOINT}/json/detail.php?apikey=abc123&id=1",
        status=200,
        body=load_fixture("detail.json"),
    )
    assert await tankerkoenig_client.station_details(station_id="1") == snapshot


async def test_nearby_stations(
    responses: aioresponses,
    tankerkoenig_client: Tankerkoenig,
    snapshot: SnapshotAssertion,
) -> None:
    """Test retrieving nearby stations."""
    responses.get(
        f"{TANKERKOENIG_ENDPOINT}/json/list.php?apikey=abc123&lat=53.1&lng=7.8&rad=5&sort=dist&type=all",
        status=200,
        body=load_fixture("list.json"),
    )
    assert (
        await tankerkoenig_client.nearby_stations(
            coordinates=(53.1, 7.8),
            radius=5,
            gas_type=GasType.ALL,
            sort=Sort.DISTANCE,
        )
        == snapshot
    )


async def test_prices(
    responses: aioresponses,
    tankerkoenig_client: Tankerkoenig,
    snapshot: SnapshotAssertion,
) -> None:
    """Test retrieving nearby stations."""
    responses.get(
        f"{TANKERKOENIG_ENDPOINT}/json/prices.php?apikey=abc123&ids=4fd02fd5-0a4c-46b3-a9b9-a488a62454be",
        status=200,
        body=load_fixture("prices.json"),
    )
    assert (
        await tankerkoenig_client.prices(
            station_ids=["4fd02fd5-0a4c-46b3-a9b9-a488a62454be"],
        )
        == snapshot
    )
