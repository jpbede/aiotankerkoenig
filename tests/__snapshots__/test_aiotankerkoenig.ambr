# serializer version: 1
# name: test_nearby_stations
  list([
    Station(id='4fd02fd5-0a4c-46b3-a9b9-a488a62454be', name='AVIA Tankstelle', house_number='115', place='Apen', post_code=26689, street='Hauptstra\\u00dfe', lat=53.22333, lng=7.812575, brand='AVIA', diesel=1.669, e10=1.689, e5=1.749, is_open=True, overrides=None, whole_day=None, state=None, opening_times=None, distance=0.9),
    Station(id='0f20f3c7-4ee0-4a9a-7d36-39d309a6ca94', name='Heinrich Albers OHG', house_number='455', place='Apen-Augustfehn', post_code=26689, street='Hauptstr.', lat=53.21672, lng=7.76265, brand='', diesel=1.659, e10=1.679, e5=1.739, is_open=True, overrides=None, whole_day=None, state=None, opening_times=None, distance=2.5),
    Station(id='5a27e942-38b4-5e83-adb8-cd396c99e0ab', name='Apen', house_number='8', place='Apen', post_code=26689, street='Dampfhammerstr.', lat=53.24424362183, lng=7.77206611633, brand='Hoyer', diesel=1.649, e10=None, e5=None, is_open=True, overrides=None, whole_day=None, state=None, opening_times=None, distance=3.3),
  ])
# ---
# name: test_prices
  dict({
    '4fd02fd5-0a4c-46b3-a9b9-a488a62454be': PriceInfo(status=<Status.OPEN: 'open'>, e5=1.739, e10=1.679, diesel=1.669),
  })
# ---
# name: test_station_details
  Station(id='abc123-0a4c-12cc-a9b9-a488a62454be', name='AVIA Tankstelle', house_number='115', place='Exampletown', post_code=26689, street='Mainstreet', lat=53.1, lng=7.8, brand='AVIA', diesel=1.669, e10=1.689, e5=1.749, is_open=True, overrides=[], whole_day=False, state=None, opening_times=None, distance=None)
# ---
